### PR TITLE
refactor(tool): wasm-opt is a tool variant

### LINKED ISSUE #685: refactor wasm-opt support to use new refactor
https://github.com/rustwasm/wasm-pack/pull/625

### PR DESCRIPTION
**Pull request by ashleygwilliams**
fixes #685 #781 #782

### CODE REVIEW

File: src/wasm_opt.rs [hunk 1]
```diff
@@ -48,61 +47,22 @@ pub fn run(
...
-            }
-        }
-    };
-
-    Ok(WasmOpt::Found(dl.binary("wasm-opt")?))
+pub fn find_wasm_opt(
+    cache: &Cache,
+    install_permitted: bool,
+) -> Result<install::Status, failure::Error> {
+    let version = "version_78";
```
Comment (EverlastingBugstopper) [inline src/wasm_opt.rs]: i see this same version number elsewhere

File: src/command/test.rs [hunk 2]
```diff
@@ -268,14 +268,17 @@ impl Test {
...
             Tool::WasmBindgen,
             &self.cache,
             &bindgen_version,
             self.mode.install_permitted(),
         )?;
 
-        self.test_runner_path = Some(dl.binary("wasm-bindgen-test-runner")?);
+        self.test_runner_path = match status {
+            install::Status::Found(dl) => Some(dl.binary("wasm-bindgen-test-runner")?),
+            _ => bail!("bleh!!"),
```
[the remark this answers is not in the document]
Comment (ashleygwilliams) [reply src/command/test.rs]: no

ok yes

File: src/install/mod.rs [hunk 3]
```diff
@@ -21,6 +22,27 @@ mod tool;
 pub use self::mode::InstallMode;
 pub use self::tool::Tool;
 
+/// Possible outcomes of attempting to find/install a tool
+pub enum Status {
+    /// Couldn't install wasm-opt because downloads are forbidden
```
[the remark this answers is not in the document]
Comment (ashleygwilliams) [reply src/install/mod.rs]: nope, that's legacy i should clean that up

File: src/install/mod.rs [hunk 4]
```diff
@@ -114,17 +137,24 @@ pub fn download_prebuilt(
...
             let binaries = &["cargo-generate"];
             match cache.download(install_permitted, "cargo-generate", binaries, &url)? {
-                Some(download) => Ok(download),
+                Some(download) => Ok(Status::Found(download)),
                 None => bail!("cargo-generate v{} is not installed!", version),
             }
         }
+        Tool::WasmOpt => {
+            let binaries = &["wasm-opt"];
+            match cache.download(install_permitted, "wasm-opt", binaries, &url)? {
```
[the remark this answers is not in the document]
Comment (ashleygwilliams) [reply src/install/mod.rs]: good call

File: src/install/mod.rs [hunk 5]
```diff
@@ -155,6 +185,13 @@ fn prebuilt_url(tool: &Tool, version: &str) -> Result<String, failure::Error> {
                 Krate::new(&Tool::CargoGenerate)?.max_version,
                 target
             ))
+        },
+        Tool::WasmOpt => {
+            Ok(format!(
+        "https://github.com/WebAssembly/binaryen/releases/download/{vers}/binaryen-{vers}-{target}.tar.gz",
+        vers = "version_78",
```
[the remark this answers is not in the document]
Comment (ashleygwilliams) [reply src/install/mod.rs]: i'll figure that out- was planning on bumping in another PR for record's sake. but if we have redundancy i should clean that up

File: src/install/tool.rs [hunk 6]
```diff
@@ -6,13 +6,17 @@ pub enum Tool {
     CargoGenerate,
     /// wasm-bindgen CLI tools
     WasmBindgen,
+    /// wasm-opt CLI tool
```
[the remark this answers is not in the document]
Comment (ashleygwilliams) [reply src/install/tool.rs]: you mean you can't tell????

no there is no reason

File: src/generate.rs [hunk 7]
```diff
@@ -1,15 +1,20 @@
...
 /// project from a template
-pub fn generate(template: &str, name: &str, download: &Download) -> Result<(), failure::Error> {
-    let bin_path = download.binary("cargo-generate")?;
+pub fn generate(
+    template: &str,
+    name: &str,
+    install_status: &install::Status,
+) -> Result<(), failure::Error> {
+    let bin_path = install::get_tool_path(install_status, Tool::CargoGenerate)?
+        .binary(&Tool::CargoGenerate.to_string())?;
```
[the remark this answers is not in the document]
Comment (ashleygwilliams) [reply src/generate.rs]: oh i should do this where the other strings are. good reminder.

[the remark this answers is not in the document]
Comment (ashleygwilliams) [reply src/generate.rs]: heh this actually turned into a much bigger rabbit hole than i expected.

File: src/install/mod.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ashleygwilliams) [reply src/install/mod.rs]: this became a similar rabbit hole to above- i think we make the enum do the work for us here but i'd like to do that work in another PR

File: src/install/tool.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (ashleygwilliams) [reply src/install/tool.rs]: oh i lied. there is a reason. there are 2 binaries in wasm-bindgen (bindgen itself and the test runner. wasm-opt is a single binary)

File: src/install/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (ashleygwilliams) [reply src/install/mod.rs]: i'm gonna do this in a separate PR

**Pull request merged**